<a href="https://colab.research.google.com/github/kepin7/Repository_DKA/blob/main/jurnal_mod1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div align="center">
<h1>Jurnal Praktikum Dasar Kecerdasan Artifisial</h1>
<h3>Modul 1: Pengenalan Python</h3>
</div>

### Tujuan Praktikum
- Mahasiswa mengetahui instalasi dan *working environment* pada Python (menggunakan Terminal dan *Jupyter Notebook*)
- Mahasiswa memahamai cara pembuatan program dengan Python
- Mahasiswa mengetahui fitur-fitur utama yang tersedia pada Python

### 1. Program Sederhana dengan Python (Bobot Nilai: `5`)

**a.** Buat program yang menampilkan kata "Hello, World!" pada terminal dengan menggunakan `print()`

In [ ]:
print("Hello, World!") # Ganti nilai None dengan nilai yang sesuai

Hello, World!


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |'Hello, World!'|

**b.** Isi variabel rasio dengan akar dari 9 $(\sqrt{9})$.
Tip: untuk menghitung akar pangkat 2 dari $x$, dapat menggunakan x**(1/2)

In [ ]:
akar_sembilan = 9**(1/2) # Ganti nilai None dengan nilai yang sesuai
print(akar_sembilan)

3.0


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |'3.0'|

**c.** Program di bawah ini menghitung nilai dari persamaan kuadrat $20x^2 + 25x + 30$ dengan meminta *input* dari pengguan untuk variabel $x$. Minta *input* pada pengguna dengan deskripsi `"Masukkan nilai x: "`, kemudian ubah variabel `x` menjadi `float`

In [ ]:
x = input("Masukkan nilai x: ") # Minta input user dengan deskripsi yang sesuai
x = float(x) # Ubah tipe data x menjadi float

hasil_persamaan = 20*x**2 + 25*x + 30
print(hasil_persamaan)

Masukkan nilai x: 3
285.0


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan nilai x: `3` |'285'|
|Masukkan nilai x: `4` |'450'|
|Masukkan nilai x: `5`|'655'|

### 2.  Pengkondisian dalam Python (Bobot Nilai: `10`)

**a.** Buat program untuk mengecek apakah suatu bilangan ganjil atau genap. Jika bilangan tersebut ganjil maka tampilkan `"Bilangan Ganjil"` pada terminal, sedangkan jika bilangan tersebut genap maka tampilkan `"Bilangan Genap"` pada terminal. Minta *input* pada pengguna dengan deskripsi (tidak ada ketentuan khusus untuk deskripsi), kemudian ubah variabel yang menampung hasil *input* pengguna menjadi `.integer`

In [ ]:
n = input("Masukkan nilai: ") # Minta input user dengan deskripsi yang sesuai
n = int(n) # Ubah tipe data n menjadi integer

if n % 2 == 0 : # Cek apakah x nilai genap
    print("Bilangan Genap")
else:
    print("Bilangan Ganjil")

Masukkan nilai: 3
Bilangan Ganjil


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan nilai: `3` |'Bilangan Ganjil'|
|Masukkan nilai: `4` |'Bilangan Genap'|
|Masukkan nilai: `5`|'Bilangan Ganjil'|

**b.** Buat program untuk meminta *input* nilai dari pengguna dalam rentang 0 hingga 100. Program tersebut menerima nilai dan menampilkan indeks huruf dari nilai tersebut pada terminal sesuai dengan kriteria berikut
- **A**: 85 - 100
- **B**: 70 - 84
- **C**: 55 - 69
- **D**: 40 - 54
- **E**: 0 - 39

In [ ]:
nilai = input("Masukkan nilai user: ") # Minta input user dengan deskripsi yang sesuai
nilai = float(nilai) # Ubah tipe data nilai menjadi float

# Buat program di atas dengan menggunakan pengkondisian
if nilai >= 85 and nilai <= 100:
    print("A")
elif nilai >= 70 and nilai <= 84:
    print("B")
elif nilai >= 55 and nilai <= 69:
    print("C")
elif nilai >= 40 and nilai <= 54:
    print("D")
elif nilai >= 0 and nilai <= 39:
    print("E")
else:
    print("Nilai tidak valid")

Masukkan nilai user: 91
A


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan nilai user: `91` |'A'|
|Masukkan nilai user: `75` |'B'|
|Masukkan nilai user: `42`|'D'|

### 3. Prosedur dan Fungsi (Bobot Nilai: `10`)

**a.** Buatlah sebuah prosedur yang menerima parameter berupa nama. Jika nama yang diberikan adalah `"Reza"`, prosedur akan mencetak output `"Selamat datang kembali, [nama]"`. Jika nama yang diberikan bukan "Reza", prosedur akan mencetak output `"Selamat datang, [nama]"`. Sesuaikan output berdasarkan nama yang diterima oleh parameter.

In [ ]:
def greet_reza(nama):
    if nama == "Reza":
        print("Selamat datang kembali, " + nama)
    else:
        print("Selamat datang, " + nama)

input_nama = input("Masukkan nama: ")  # Minta input user dengan deskripsi yang sesuai

greet_reza(input_nama)  # Panggil prosedur `greet_reza()` dengan parameter yang sesuai

Masukkan nama: Reza
Selamat datang kembali, Reza


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan nama: `Afeef` |'Selamat datang, Afeef'|
|Masukkan nama: `Reza` |'Selamat datang kembali, Reza'|
|Masukkan nama: `Tiara`|'Selamat datang, Tiara'|

**b.** Buatlah sebuah fungsi sederhana yang mengonversi jarak dari kilometer (KM) menjadi meter (M). Fungsi ini akan menerima input berupa jarak dalam kilometer, kemudian mengembalikan hasil konversi dalam meter.

In [ ]:
def km_ke_m(km): # Nama fungsi dibebaskan

    return km*1000 # Kembalikan hasil perhitungan konversi

jarak_km = input("Masukkan jarak (KM): ") # Minta input user dengan deskripsi yang sesuai
jarak_km = float(jarak_km) # Ubah tipe data input dari user menjadi tipe data yang sesuai (lihat test case)

hasil_konversi = km_ke_m(jarak_km) # Panggil fungsi yang sudah dibuat dan hasilnya disimpan ke dalam variabel `hasil_konversi`
print(hasil_konversi) # Menampilkan hasil konversi ke terminal

Masukkan jarak (KM): 5.5
5500.0


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan jarak (KM): `3` |'3000'|
|Masukkan jarak (KM): `4` |'4000'|
|Masukkan jarak (KM): `5.5`|'5500'|

### 4. Perulangan (Bobot Nilai: `15`)

**a.** Buatlah program yang menerima *input* dari pengguna berupa sebuah bilangan bulat positif `n`. Program tersebut harus menampiklan semua bilangan ganjil dari 1 hingga `n` menggunakan perulangan `for` dan fungsi `range()`. Top: gunakan langkah kelipatan 2 dalam `range()` untuk menampilkan hanya angka ganjil.

In [16]:
# Buat kode untuk meminta input pengguna dan konversi input ke integer
n = int(input("Masukkan n: "))

# Perulangan untuk menampilkan angka ganjil sesuai input user
for i in range(1, n+1, 2) : # isi nilai none
    print(i)

Masukkan n: 11
1
3
5
7
9
11


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan n: `1` |'1'|
|Masukkan n: `6` |'1'<br>'1'<br>'5'|
|Masukkan n: `11`|'1'<br>'3'<br>'5'<br>'7'<br>'9'<br>'11'|

**b.** Buatlah sebuah fungsi yang menerima argumen `n` dan mengembalikan jumlah dari bilangan bulat dari 1 hingga n.  Sebagai contoh, jika `n = 5`, fungsi tersebut harus menghitung penjumlahan **$1+2+3+4+5$** dan mengembalikan nilai 15. Jika `n` kurang dari 1, kembalikan nilai 0. Gunakan perulangan `while` untuk menghitung jumlah tersebut.

In [19]:
def hitung_n(n) : # Buat fungsi untuk menghitung penjumlahan
    if n<1 : # Kondisi jika n kurang dari 1
        return 0

    else :
        total = 0
        i = 1
        while i <= n: # Buat kondisi untuk melakukan perulangan dari 1 hingga n
            total += i # Tambah total dengan i
            i += 1 # Tambah i dengan 1 sebagai iterator

        return total # Mengembalikan variabel yang menampung nilai total


# Buat kode untuk meminta input pengguna dan konversi input ke integer
n = int(input("Masukkan n: "))

hasil = hitung_n(n) # Panggil fungsi untuk menghitung penjumlahan dan simpan hasilnya kedalam variabel
print(hasil) # Tampilkan hasil penjumlahan yang sudah dihitung ke terminal

Masukkan n: 10
55


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|Masukkan n: `5` |'15'|
|Masukkan n: `10` |'55'|
|Masukkan n: `-3`|'0'|

### 5. List dan Dictionary (Bobot Nilai: `20`)

**a.** Buat program yang mengganti ekstensi `.go` menjadi `.py` pada setiap elemen dalam list `filenames`. Gunakan metode `.replace()` untuk mengubah ekstensi, kemudian simpan nama file yang telah diubah ke dalam list baru bernama `new_filenames` dan kemudian cetak list ke dalam terminal.

Tip :
```python
sapa_tiara = "Hai Tiara!!"
print("Sapa Tiara:", sapa_tiara)

sapa_farah = sapa_tiara.replace("Tiara", "Farah")
print("Sapa Farah:", sapa_farah)
```

output
```
Sapa Tiara: Hai Tiara!!
Sapa Farah: Hai Farah!!
```

In [21]:
filenames = ["app.go", "main.go", "formula.go", "program.go"]
new_filenames = [] # Buat list kosong

for filename in filenames : # Buat loop untuk iterasi elemen pada lit filenames
    new_name = filename.replace(".go", ".py") # Ganti ekstensi .go menjadi .py menggunakan .replace()
    new_filenames.append(new_name) # Masukkan nama yang sudah diganti ke dalam list new_filenames

# Cetak list yang berisi nama file yang sudah diganti
print(new_filenames)

['app.py', 'main.py', 'formula.py', 'program.py']


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |['app.py', 'main.py', 'formula.py', 'program.py']|

Buatlah fungsi `email_list` yang menerima sebuah dictionary. Dictonary ini berisi nama domain sebagai `key`, dan daftar pengguna sebagai `value`. Lengkapi bagian yang kosong dalam fungsi untuk menghasilkan daftar yang berisi alamat email lengkap (misalnya, `dwiva.arya@gmail.com`). Alamat email ini dihasilkan dengan menggabungkan setiap nama pengguna dengan nama domain yang sesuai.

Tip:
```python
gender_dict = {
    "Rafly" : "Laki-Laki",
    "Iksan" : "Laki-Laki",
    "Imelda" : "Perempuan"
}

for key, value in gender_dict.items() :
    print("Key:", key, ", Value:", value)
```

Output
```
Key: Rafly , Value: Laki-Laki
Key: Iksan , Value: Laki-Laki
Key: Imelda , Value: Perempuan
```

In [24]:
def email_list(domains) :
    emails = [] # List untuk menampung email lengkap

    # Loop untuk setiap key dan value pada dict domains
    for domains, users in domains.items() :
        for user in users:
            # Tambahakan email lengkap ke list emails
            emails.append(user+'@'+domains)

    return emails

email_domain = {
    "gmail.com": ["rafly.arj", "syahreza.adnan", "valen.hartanto"],
    "telkomuniversity.ac.id" : ["sabrina.tiara", "imelda"],
    "yahoo.com" : ["raihan.omen", "dito.asyraf"]
}

# Panggil fungsi email_list, kemudian masukkan ke dalam variabel
hasil = print(email_list(email_domain))

print(hasil) # Cetak list hasil email lengkap yang didaptkan dari email_list

['rafly.arj@gmail.com', 'syahreza.adnan@gmail.com', 'valen.hartanto@gmail.com', 'sabrina.tiara@telkomuniversity.ac.id', 'imelda@telkomuniversity.ac.id', 'raihan.omen@yahoo.com', 'dito.asyraf@yahoo.com']
None


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |['rafly.arj@gmail.com', 'syahreza.adnan@gmail.com', 'valen.hartanto@gmail.com', 'sabrina.tiara@telkomuniversity.ac.id', 'imelda@telkomuniversity.ac.id', 'raihan.omen@yahoo.com', 'dito.asyraf@yahoo.com']|

### 6. Soal Cerita (Bobot Nilai: `40`)

Seorang guru ingin membuat program sederhana untuk mencatat nilai ujian siswa. Setiap siswa akan mengikuti **tiga mata pelajaran**. Guru perlu menghitung rata-rata nilai untuk setiap siswa dan menentukan apakah siswa tersebut **lulus** atau **tidak**. Kriteria kelulusan adalah jika rata-rata nilai siswa sama dengan atau lebih dair **70**.

Buatlah program yang melakukan hal berikut :
1. **Input/Output**
    - Meminta input dari pengguna berupa `nama siswa` dan `nilai` untuk tiga mata pelajaran
    - Menampilkan **rata-rata nilai** dan **status kelulusan** setiap siswa

2. **Penentuan kelulusan**
    - Menentukan apakah siswa **lulus** atau **tidak** berdasarkan rata-rata nilainya.

3. **Pengulangan program**
    - Menggunakan perulangan untuk memasukkan data beberapa siswa hingga pengguna memasukkan `"selesai"` sebagai nama siswa

4. **Perhitungan rata-rata**
    - Membuat fungsi untuk **menghitung rata-rata** yang menerima list nilai sebagai parameter dan mengembalikan rata-rata nilai dari list tersebut

5. **Penyimpanan data**
    - Gunakan *dictionary* dengan `keys` nama siswa, dan `value` list nilai dari siswa tersebut.

In [26]:
# You’re on Your Own, Kid. Good Luck :)
# Fungsi untuk menghitung rata-rata nilai
def hitung_rata_rata(nilai_list):
    return sum(nilai_list) / len(nilai_list)

# Dictionary untuk menyimpan data siswa
data_siswa = {}

# Pengulangan program
while True:
    # Input nama siswa
    nama = input("Masukkan nama siswa (atau ketik 'selesai' untuk berhenti): ")

    # Cek apakah pengguna ingin berhenti
    if nama.lower() == 'selesai':
        break

    # List untuk menyimpan nilai
    nilai_list = []

    # Input nilai untuk tiga mata pelajaran
    for i in range(1, 4):
        nilai = int(input(f"Masukkan nilai mata pelajaran {i}: "))
        nilai_list.append(nilai)

    # Hitung rata-rata
    rata_rata = hitung_rata_rata(nilai_list)

    # Tentukan status kelulusan
    status = "Lulus" if rata_rata >= 70 else "Tidak Lulus"

    # Tampilkan hasil
    print(f"Rata-rata nilai {nama} adalah {rata_rata:.2f}. Status: {status}")

    # Simpan data siswa
    data_siswa[nama] = {
        'nilai': nilai_list,
        'rata_rata': rata_rata,
        'status': status
    }

# Tampilkan seluruh data siswa
print("\nData Siswa:")
for nama, info in data_siswa.items():
    print(f"- {nama}: {info['nilai']}, Rata-rata: {info['rata_rata']:.2f}, Status: {info['status']}")

Masukkan nama siswa (atau ketik 'selesai' untuk berhenti): Andi
Masukkan nilai mata pelajaran 1: 80
Masukkan nilai mata pelajaran 2: 75
Masukkan nilai mata pelajaran 3: 90
Rata-rata nilai Andi adalah 81.67. Status: Lulus
Masukkan nama siswa (atau ketik 'selesai' untuk berhenti): selesai

Data Siswa:
- Andi: [80, 75, 90], Rata-rata: 81.67, Status: Lulus


**Bagian yang diberikan tanda `ini` adalah masukkan user**

*Test case* program

**|----------------------------------------------------------------------------------|**

Masukkan nama siswa (atau ketik 'selesai' untuk berhenti): `Andi`

Masukkan nilai mata pelajaran 1: `80`

Masukkan nilai mata pelajaran 2: `75`

Masukkan nilai mata pelajaran 3: `90`

Rata-rata nilai Andi adalah 81.67. Status: Lulus

Masukkan nama siswa (atau ketik 'selesai' untuk berhenti): `Budi`

Masukkan nilai mata pelajaran 1: `60`

Masukkan nilai mata pelajaran 2: `65`

Masukkan nilai mata pelajaran 3: `70`

Rata-rata nilai Budi adalah 65.00. Status: Tidak Lulus

Masukkan nama siswa (atau ketik 'selesai' untuk berhenti): `selesai`

Data Siswa:
- Andi: [80, 75, 90], Rata-rata: 81.67, Status: Lulus
- Budi: [60, 65, 70], Rata-rata: 65.00, Status: Tidak Lulus

**|----------------------------------------------------------------------------------|**